src/ucp/rma/basic_rma.c

```diff
@@ -24,86 +24,107 @@
 {
-    ucp_ep_rma_config_t *rma_config;
-    ucs_status_t status;
-    uct_rkey_t uct_rkey;
-    size_t frag_length;
-    ssize_t packed_len;
-    ucp_lane_index_t lane;
+    return length > rma_config->max_put_short && length > 16384;
+}
 
-    UCP_RMA_CHECK_PARAMS(buffer, length);
+static inline int ucp_rma_get_is_zcopy(ucp_ep_rma_config_t *rma_config, size_t length)
+{
+    return length > 16384;
+}
 
-    /* Loop until all message has been sent.
-     * We re-check the configuration on every iteration, because it can be
-     * changed by transport switch.
+static inline void ucp_rma_request_wait(ucp_request_t *req)
+{
+    while(req->send.uct_comp.count > 0) {
+        ucp_worker_progress(req->send.ep->worker);
+    }
+}
+
+static void ucp_rma_request_tryfree(ucp_request_t *req)
+{
+    /* request can be released if 
+     *  - all fragments were sent and
+     *  - all zcopy fragments are done
+     *  - and request was allocated from the mpool
      */
-    for (;;) {
-        UCP_EP_RESOLVE_RKEY_RMA(ep, rkey, lane, uct_rkey, rma_config);
-        if (length <= rma_config->max_put_short) {
-            status = uct_ep_put_short(ep->uct_eps[lane], buffer, length,
-                                      remote_addr, uct_rkey);
-            if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
-                break;
-            }
-        } else {
-            if (length <= ucp_ep_config(ep)->bcopy_thresh) {
-                frag_length = ucs_min(length, rma_config->max_put_short);
-                status = uct_ep_put_short(ep->uct_eps[lane], buffer, frag_length,
-                                          remote_addr, uct_rkey);
-            } else {
-                ucp_memcpy_pack_context_t pack_ctx;
-                pack_ctx.src    = buffer;
-                pack_ctx.length = frag_length =
-                                ucs_min(length, rma_config->max_put_bcopy);
-                packed_len = uct_ep_put_bcopy(ep->uct_eps[lane], ucp_memcpy_pack,
-                                              &pack_ctx, remote_addr, uct_rkey);
-                status = (packed_len > 0) ? UCS_OK : (ucs_status_t)packed_len;
-            }
-            if (ucs_likely(status == UCS_OK)) {
-                length      -= frag_length;
-                if (length == 0) {
-                    break;
-                }
-
-                buffer      += frag_length;
-                remote_addr += frag_length;
-            } else if (status != UCS_ERR_NO_RESOURCE) {
-                break;
-            }
-        }
-        ucp_worker_progress(ep->worker);
+   if ((req->send.length == 0) && 
+       (req->send.uct_comp.count == 0) && 
+       (req->flags & UCP_REQUEST_FLAG_RELEASED)) {
+
+        ucp_request_put(req, UCS_OK);
+   }
+}
+
+static void ucp_rma_request_bcopy_completion(uct_completion_t *self, ucs_status_t status)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct_comp);
+
+    if (status == UCS_ERR_NO_RESOURCE) {
```

Comment:
we can but it will be the same if in both progress_put_inner and progress_get_inner